### DISCUSSION TITLE
Either Zeptoedit/edit OR REPL ==> Zeptoedit/Edit with REPL

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#117] [Ideas]
Hi,
(this is nothing Moore :) than an idea. I am just starting with Zeptoforth I am already totally fascinated by Zeptoforth and by 
the "how Forth does it"-thing.

One extraodinary cool feature of Forth is the REPL and Zeptoforth ability to circumvent a restriction, which many other FORTHs have: To experiment with words like IF ELSE THEN, DO LOOP and similiar via the REPL, which could only be used inside word definitions normally. Zeptoforth does implicit compiling! Yeah! :)

So, imagine one has come up with a very cool idea what to program next, s/he tries some things on the REPL...ok....start
"10 edit"....write some code...and HECK!...how was the stack parameter sequence for the word xyz???...back to the REPL:
Save block...(flash write cycle engaged....)
Try the word at the REPL...AH! **Enlightenment!**...
10 edit...again
restore the cursor position again and start programming.

...

The idea is to have a key shortcut, which opens A REPL "below" the "block field" of edit/zed which expands to the top with each command/result makeing the "block field" of edit/zed smaller.
Leaving the REPL by a certain shortcut (may be CTR-D) will end the REPL and resize the "block field" to its full glory putting
the cursor where it was left before.

The whole thing could be - user point of view only - like leaving n/vim with CTRL-Z and entering again with fg from the commandline. No integration of a REPL _into_ edit or zed needed, just putting edit/zed "into the background" and kind of freezing its state when jumping to the REPL.

Cheers!
Tuxic

**Comment by tabemann** [comment]
On Sun, Jul 21, 2024, 10:09 mcctuxic ***@***.***> wrote:

> Hi,
> (this is nothing Moore :) than an idea. I am just starting with Zeptoforth
> I am already totally fascinated by Zeptoforth and by
> the "how Forth does it"-thing.
>
> One extraodinary cool feature of Forth is the REPL and Zeptoforth ability
> to circumvent a restriction, which many other FORTHs have: To experiment
> with words like IF ELSE THEN, DO LOOP and similiar via the REPL, which
> could only be used inside word definitions normally. Zeptoforth does
> implicit compiling! Yeah! :)
>
> So, imagine one has come up with a very cool idea what to program next,
> s/he tries some things on the REPL...ok....start
> "10 edit"....write some code...and HECK!...how was the stack parameter
> sequence for the word xyz???...back to the REPL:
> Save block...(flash write cycle engaged....)
> Try the word at the REPL...AH! *Enlightenment!*...
> 10 edit...again
> restore the cursor position again and start programming.
>
> ...
>
> The idea is to have a key shortcut, which opens A REPL "below" the "block
> field" of edit/zed which expands to the top with each command/result
> makeing the "block field" of edit/zed smaller.
> Leaving the REPL by a certain shortcut (may be CTR-D) will end the REPL
> and resize the "block field" to its full glory putting
> the cursor where it was left before.
>
> The whole thing could be - user point of view only - like leaving n/vim
> with CTRL-Z and entering again with fg from the commandline. No integration
> of a REPL *into* edit or zed needed, just putting edit/zed "into the
> background" and kind of freezing its state when jumping to the REPL.
>
> Cheers!
> Tuxic
>

I thought of doing just this but decided against it due to a fundamental
architectural limitation of zeptoforth - memory space alloted by, say,
zeptoed acts like a stack - zeptoed frees memory alloted to itself by
restoring the main task's RAM dictionary pointer to its original state when
it exits. However, if one compiled code to RAM while running zeptoed either
zeptoed would not be able to free its space or all that code would have to
be forgotten.

Additionally if I were to implement some split-screen arrangement I would
have to implement functionality for intercepting all the output from Forth
code other than zeptoed or edit, converting it to be displayed in the split
screen, and intercepting ANSI codes as needed (and generating replies in
some cases).

Fundamentally, it sounds like a really neat idea but gets bogged down in
details when you actually think of implementing it.

Travis

**Comment by tabemann** [comment]
On Sun, Jul 21, 2024 at 12:34 PM Travis Bemann ***@***.***> wrote:

> On Sun, Jul 21, 2024, 10:09 mcctuxic ***@***.***> wrote:
>
>> Hi,
>> (this is nothing Moore :) than an idea. I am just starting with
>> Zeptoforth I am already totally fascinated by Zeptoforth and by
>> the "how Forth does it"-thing.
>>
>> One extraodinary cool feature of Forth is the REPL and Zeptoforth ability
>> to circumvent a restriction, which many other FORTHs have: To experiment
>> with words like IF ELSE THEN, DO LOOP and similiar via the REPL, which
>> could only be used inside word definitions normally. Zeptoforth does
>> implicit compiling! Yeah! :)
>>
>> So, imagine one has come up with a very cool idea what to program next,
>> s/he tries some things on the REPL...ok....start
>> "10 edit"....write some code...and HECK!...how was the stack parameter
>> sequence for the word xyz???...back to the REPL:
>> Save block...(flash write cycle engaged....)
>> Try the word at the REPL...AH! *Enlightenment!*...
>> 10 edit...again
>> restore the cursor position again and start programming.
>>
>> ...
>>
>> The idea is to have a key shortcut, which opens A REPL "below" the "block
>> field" of edit/zed which expands to the top with each command/result
>> makeing the "block field" of edit/zed smaller.
>> Leaving the REPL by a certain shortcut (may be CTR-D) will end the REPL
>> and resize the "block field" to its full glory putting
>> the cursor where it was left before.
>>
>> The whole thing could be - user point of view only - like leaving n/vim
>> with CTRL-Z and entering again with fg from the commandline. No integration
>> of a REPL *into* edit or zed needed, just putting edit/zed "into the
>> background" and kind of freezing its state when jumping to the REPL.
>>
>> Cheers!
>> Tuxic
>>
>
> I thought of doing just this but decided against it due to a fundamental
> architectural limitation of zeptoforth - memory space alloted by, say,
> zeptoed acts like a stack - zeptoed frees memory alloted to itself by
> restoring the main task's RAM dictionary pointer to its original state when
> it exits. However, if one compiled code to RAM while running zeptoed either
> zeptoed would not be able to free its space or all that code would have to
> be forgotten.
>
> Additionally if I were to implement some split-screen arrangement I would
> have to implement functionality for intercepting all the output from Forth
> code other than zeptoed or edit, converting it to be displayed in the split
> screen, and intercepting ANSI codes as needed (and generating replies in
> some cases).
>
> Fundamentally, it sounds like a really neat idea but gets bogged down in
> details when you actually think of implementing it.
>

BTW, you should grab the very newest zeptoed from the zeptoforth repository
─ it includes a range of important performance optimizations which I found
were necessary when running zeptoIP simultaneously (note that when running
zeptoed and zeptoIP at the same time you will need to shrink the zeptoed
heap or otherwise it will crash). If you grabbed it last night, grab it
again because I accidentally introduced a stupid bug which I just fixed.

Travis

**Comment by tabemann** [comment]
On Sun, Jul 21, 2024 at 12:58 PM Travis Bemann ***@***.***> wrote:

>
>
> On Sun, Jul 21, 2024 at 12:34 PM Travis Bemann ***@***.***> wrote:
>
>> On Sun, Jul 21, 2024, 10:09 mcctuxic ***@***.***> wrote:
>>
>>> Hi,
>>> (this is nothing Moore :) than an idea. I am just starting with
>>> Zeptoforth I am already totally fascinated by Zeptoforth and by
>>> the "how Forth does it"-thing.
>>>
>>> One extraodinary cool feature of Forth is the REPL and Zeptoforth
>>> ability to circumvent a restriction, which many other FORTHs have: To
>>> experiment with words like IF ELSE THEN, DO LOOP and similiar via the REPL,
>>> which could only be used inside word definitions normally. Zeptoforth does
>>> implicit compiling! Yeah! :)
>>>
>>> So, imagine one has come up with a very cool idea what to program next,
>>> s/he tries some things on the REPL...ok....start
>>> "10 edit"....write some code...and HECK!...how was the stack parameter
>>> sequence for the word xyz???...back to the REPL:
>>> Save block...(flash write cycle engaged....)
>>> Try the word at the REPL...AH! *Enlightenment!*...
>>> 10 edit...again
>>> restore the cursor position again and start programming.
>>>
>>> ...
>>>
>>> The idea is to have a key shortcut, which opens A REPL "below" the
>>> "block field" of edit/zed which expands to the top with each command/result
>>> makeing the "block field" of edit/zed smaller.
>>> Leaving the REPL by a certain shortcut (may be CTR-D) will end the REPL
>>> and resize the "block field" to its full glory putting
>>> the cursor where it was left before.
>>>
>>> The whole thing could be - user point of view only - like leaving n/vim
>>> with CTRL-Z and entering again with fg from the commandline. No integration
>>> of a REPL *into* edit or zed needed, just putting edit/zed "into the
>>> background" and kind of freezing its state when jumping to the REPL.
>>>
>>> Cheers!
>>> Tuxic
>>>
>>
>> I thought of doing just this but decided against it due to a fundamental
>> architectural limitation of zeptoforth - memory space alloted by, say,
>> zeptoed acts like a stack - zeptoed frees memory alloted to itself by
>> restoring the main task's RAM dictionary pointer to its original state when
>> it exits. However, if one compiled code to RAM while running zeptoed either
>> zeptoed would not be able to free its space or all that code would have to
>> be forgotten.
>>
>> Additionally if I were to implement some split-screen arrangement I would
>> have to implement functionality for intercepting all the output from Forth
>> code other than zeptoed or edit, converting it to be displayed in the split
>> screen, and intercepting ANSI codes as needed (and generating replies in
>> some cases).
>>
>> Fundamentally, it sounds like a really neat idea but gets bogged down in
>> details when you actually think of implementing it.
>>
>
> BTW, you should grab the very newest zeptoed from the zeptoforth
> repository ─ it includes a range of important performance optimizations
> which I found were necessary when running zeptoIP simultaneously (note that
> when running zeptoed and zeptoIP at the same time you will need to shrink
> the zeptoed heap or otherwise it will crash). If you grabbed it last night,
> grab it again because I accidentally introduced a stupid bug which I just
> fixed.
>

And if you grabbed the latest zeptoed before, I suggest you grab it again,
as I made a number of important bugfixes to it later yesterday.

Travis

**Comment by mcctuxic** [comment]
On 07/22 07:29, tabemann wrote:
> On Sun, Jul 21, 2024 at 12:58 PM Travis Bemann ***@***.***> wrote:
> 
> >
> >
> > On Sun, Jul 21, 2024 at 12:34 PM Travis Bemann ***@***.***> wrote:
> >
> >> On Sun, Jul 21, 2024, 10:09 mcctuxic ***@***.***> wrote:
> >>
> >>> Hi,
> >>> (this is nothing Moore :) than an idea. I am just starting with
> >>> Zeptoforth I am already totally fascinated by Zeptoforth and by
> >>> the "how Forth does it"-thing.
> >>>
> >>> One extraodinary cool feature of Forth is the REPL and Zeptoforth
> >>> ability to circumvent a restriction, which many other FORTHs have: To
> >>> experiment with words like IF ELSE THEN, DO LOOP and similiar via the REPL,
> >>> which could only be used inside word definitions normally. Zeptoforth does
> >>> implicit compiling! Yeah! :)
> >>>
> >>> So, imagine one has come up with a very cool idea what to program next,
> >>> s/he tries some things on the REPL...ok....start
> >>> "10 edit"....write some code...and HECK!...how was the stack parameter
> >>> sequence for the word xyz???...back to the REPL:
> >>> Save block...(flash write cycle engaged....)
> >>> Try the word at the REPL...AH! *Enlightenment!*...
> >>> 10 edit...again
> >>> restore the cursor position again and start programming.
> >>>
> >>> ...
> >>>
> >>> The idea is to have a key shortcut, which opens A REPL "below" the
> >>> "block field" of edit/zed which expands to the top with each command/result
> >>> makeing the "block field" of edit/zed smaller.
> >>> Leaving the REPL by a certain shortcut (may be CTR-D) will end the REPL
> >>> and resize the "block field" to its full glory putting
> >>> the cursor where it was left before.
> >>>
> >>> The whole thing could be - user point of view only - like leaving n/vim
> >>> with CTRL-Z and entering again with fg from the commandline. No integration
> >>> of a REPL *into* edit or zed needed, just putting edit/zed "into the
> >>> background" and kind of freezing its state when jumping to the REPL.
> >>>
> >>> Cheers!
> >>> Tuxic
> >>>
> >>
> >> I thought of doing just this but decided against it due to a fundamental
> >> architectural limitation of zeptoforth - memory space alloted by, say,
> >> zeptoed acts like a stack - zeptoed frees memory alloted to itself by
> >> restoring the main task's RAM dictionary pointer to its original state when
> >> it exits. However, if one compiled code to RAM while running zeptoed either
> >> zeptoed would not be able to free its space or all that code would have to
> >> be forgotten.
> >>
> >> Additionally if I were to implement some split-screen arrangement I would
> >> have to implement functionality for intercepting all the output from Forth
> >> code other than zeptoed or edit, converting it to be displayed in the split
> >> screen, and intercepting ANSI codes as needed (and generating replies in
> >> some cases).
> >>
> >> Fundamentally, it sounds like a really neat idea but gets bogged down in
> >> details when you actually think of implementing it.
> >>
> >
> > BTW, you should grab the very newest zeptoed from the zeptoforth
> > repository ─ it includes a range of important performance optimizations
> > which I found were necessary when running zeptoIP simultaneously (note that
> > when running zeptoed and zeptoIP at the same time you will need to shrink
> > the zeptoed heap or otherwise it will crash). If you grabbed it last night,
> > grab it again because I accidentally introduced a stupid bug which I just
> > fixed.
> >
> 
> And if you grabbed the latest zeptoed before, I suggest you grab it again,
> as I made a number of important bugfixes to it later yesterday.
> 
> Travis

I just zeptoed (verb) my current installation with the newest Zeptoed ever ! :)

Thanks for all this great stuff and your support, Travis!

Tuxic
